### PR TITLE
script: throw disabled err for op_ver and its variants

### PR DESCRIPTION
**Pull request by ChrisCho-H**
Throw `DISABLED_OPCODE` error for op_ver and its variants as it's known but just disabled opcodes. Currently, interpreter throws `BAD_OPCODE` which contains the message of  `"Opcode missing or not understood"`, which isn't accurate and confusing.

### CODE REVIEW

**Comment by achow101** [comment]
Concept ACK

I agree that `BAD_OPCODE` is the wrong error for these opcodes. However, instead of adding a new opcode, I think these should just return the existing `DISABLED_OPCODE` error that is used for the other disabled opcodes.

Also `OP_VER` should be included here as well as that's really what is disabled. `OP_VERIF` and `OP_VERNOTIF` are really just variants of `OP_VER`.

**Comment by Dearfor** [comment]
Ok thanks.

**Comment by ChrisCho-H** [comment]
Thx! I was also considering it but not 100% sure whether `OP_VERIF` and `OP_VERNOTIF` can be treated as disabled. I will use the existing one as your suggestion.

**Review by jonatack — COMMENTED** [review]
Approach ACK ffc32938aa962f7c4b8d5f4af69bd710c174ef44

The unit script tests (`./src/test/test_bitcoin -t script_tests`) don't pass at the first commit, so it may be best to combine the second commit updating the tests into the first commit with the code changes.

**Comment by ChrisCho-H** [comment]
squashed into single commit ceab6bf3ac6b09ae7f210c05614d2c898ef7ee37

**Comment by ChrisCho-H** [comment]
updated PR description

File: src/script/interpreter.cpp [hunk 1]
```diff
@@ -467,7 +467,9 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                 opcode == OP_DIV ||
                 opcode == OP_MOD ||
                 opcode == OP_LSHIFT ||
-                opcode == OP_RSHIFT)
+                opcode == OP_RSHIFT ||
+                opcode == OP_VERIF ||
+                opcode == OP_VERNOTIF)
```
Comment (achow101) [inline src/script/interpreter.cpp:472]: `OP_VER` can be included here too:

```suggestion
                opcode == OP_VERNOTIF ||
                (fExec && opcode == OP_VER))
```

Comment (ChrisCho-H) [reply src/script/interpreter.cpp]: cool, I've implemented!

File: src/script/interpreter.cpp [hunk 2]
```diff
@@ -467,7 +467,10 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                 opcode == OP_DIV ||
                 opcode == OP_MOD ||
                 opcode == OP_LSHIFT ||
-                opcode == OP_RSHIFT)
+                opcode == OP_RSHIFT ||
+                opcode == OP_VERIF ||
+                opcode == OP_VERNOTIF ||
+                (fExec && opcode == OP_VER))
```
Comment (luke-jr) [inline src/script/interpreter.cpp:473]: How does OP_VER differ from a non-existent opcode?

Comment (achow101) [reply src/script/interpreter.cpp:473]: It's actually defined and exists in documentation.

**Comment by ajtowns** [comment]
I don't understand why we'd want to touch consensus code for something that's at most a slightly confusing error message? Just changing the error description as in #28234 seems a strictly better approach.

As it stands, `SCRIPT_ERR_DISABLED_OPCODE` covers opcodes that are unusable even in unexecuted IF/ELSE branches, while `SCRIPT_ERR_BAD_OPCODE` covers opcodes that are only unusable when evaluated.

**Comment by ChrisCho-H** [comment]
@ajtowns https://github.com/bitcoin/bitcoin/pull/28234 is for reserved opcode, but this change is for disabled opcode.
Moreover, it's not just about confusing err msg but also for the efficiency which I said as a second reason above.
I agree that we need to be conservative for consensus code, though I don't see the side effect of this change(let me know if there is!).

**Comment by ajtowns** [comment]
There's no difference between "reserving" an opcode and "disabling" it other than the error message that's passed through when it's used. Having different error messages for "error even if not executed" and "error only when it's executed" might make sense, but that isn't what we currently do (OP_VERIF/OP_VERNOTIF are exceptions) and isn't what this PR currently does.

Having the error codes be:

```
        case SCRIPT_ERR_BAD_OPCODE:
            return "Attempted to execute invalid/reserved opcode";
        case SCRIPT_ERR_DISABLED_OPCODE:
            return "Disabled opcode present in script";
```

or similar, and just changing the results for VERIF/VERNOTIF by:

```c++
    ...
    default:
        if (opcode == OP_VERIF || opcode == OP_VERNOTIF) {
            return set_error(serror, SCRIPT_ERR_DISABLED_OPCODE);
        } else {
            return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
        }
```

seems simpler to ensure there's no meaningful consensus change.

> 2\. but it's such an inefficient to go through all cases even if it does not need to

A switch statement isn't made inefficient by the presence of other cases, it's just a lookup and a jump. (It is made inefficient if the cases are sparse, but they're not here)

**Comment by ChrisCho-H** [comment]
just out of curiosity, what's the difference between this(just removed `fExec && opcode == OP_VER` check)
```cpp
            if (opcode == OP_CAT ||
                opcode == OP_SUBSTR ||
                opcode == OP_LEFT ||
                opcode == OP_RIGHT ||
                opcode == OP_INVERT ||
                opcode == OP_AND ||
                opcode == OP_OR ||
                opcode == OP_XOR ||
                opcode == OP_2MUL ||
                opcode == OP_2DIV ||
                opcode == OP_MUL ||
                opcode == OP_DIV ||
                opcode == OP_MOD ||
                opcode == OP_LSHIFT ||
                opcode == OP_RSHIFT ||
                opcode == OP_VERIF ||
                opcode == OP_VERNOTIF) 
                return set_error(serror, SCRIPT_ERR_DISABLED_OPCODE); // Disabled opcodes (CVE-2010-5137).
```
and this?

```cpp
    ...
    default:
        if (opcode == OP_VERIF || opcode == OP_VERNOTIF) {
            return set_error(serror, SCRIPT_ERR_DISABLED_OPCODE);
        } else {
            return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
        }
```


and at first I didn't include `OP_VER` but @achow101 suggested to include it in disabled err(which makes sense to me).
however, if simply changing err msg(https://github.com/bitcoin/bitcoin/pull/28234) could works for `OP_VER`, we can omit check for `OP_VER`.

**Comment by ajtowns** [comment]
Changing nothing other than the `default` clause makes it easy to see that the only functional change is in the error code being returned. Changing the logic outside the `switch` requires more care to see that there aren't any other weird behaviours being introduced. That is, it's optimising for simplicity of review.

If you're wanting to give the same error message for `OP_VER` (which is only invalid if executed) the same as `OP_VERIF` (which is invalid even if not executed) then an even simpler change is to just make both `SCRIPT_ERR_BAD_OPCODE` and `SCRIPT_ERR_DISABLED_OPCODE` give the same error message and not touch consensus code at all, eg:

```c++
    case SCRIPT_ERR_BAD_OPCODE:
    case SCRIPT_ERR_DISABLED_OPCODE:
        return "Invalid opcode (disabled/reserved/unknown)";
```

**Comment by ChrisCho-H** [comment]
I think it might be better to separate `OP_VER` and its variants as functionally different(based on executed branch or not).
just throw `SCRIPT_ERR_DISABLED_OPCODE` for `OP_VERIF` and `OP_VERNOTIF`, not for `OP_VER`.
let `OP_VER` throw `SCRIPT_ERR_BAD_OPCODE` as it is.

**Comment by ChrisCho-H** [comment]
add `"reserved"` in bad opcode err message in following commit 7ec6faff12b846255395aefa2f6dfa05bcb27bd7

File: src/script/interpreter.cpp (hunk 2 above)
Comment (ChrisCho-H) [reply src/script/interpreter.cpp]: I think that's the reason why OP_VER and its variants are evaluated as disabled opcode while `OP_RESERVED` is evaluated as reserved.
actually other disabled opcodes(which throw disabled err) are based on history.

**Comment by ChrisCho-H** [comment]
I move the logic to default clause following @ajtowns opinion, but still include `OP_VER` to throw disabled error(because it's disabled opcode).
I also added `reserved` for bad opcode error in 17a0270da5536547850dca06885a278d3e0d3269 which is for reserved opcode.

I mentioned above, but If there's not any meaningful difference btw disabled and reserved ones, it would make sense to throw disabled error for both of them as it's how it works for disabled opcodes.
like below

```cpp
    ...
    default:
        if (opcode == OP_VERIF || 
            opcode == OP_VERNOTIF ||
            opcode == OP_VER ||
            opcode == OP_RESERVED ||
            opcode == OP_RESERVED1 ||
            opcode == OP_RESERVED2) {
            return set_error(serror, SCRIPT_ERR_DISABLED_OPCODE);
        } else {
            return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
        }
```

in this way we don't need to add `reserved` in bad opcode error message.

**Comment by ajtowns** [comment]
If we're not differentiating between `OP_VER` and `OP_VERIF`, I think we should just merge the `DISABLED_OPCODE` and `BAD_OPCODE` errors.

**Comment by ChrisCho-H** [comment]
If `DISABLED_OPCODE` and `BAD_OPCODE` error throw same message, do we need to have both case?

**Comment by ChrisCho-H** [comment]
I think it's better to dump 2d6530b4c5300a29d87fb7042c6cec0b08cc9a0f. and reset to 17a0270da5536547850dca06885a278d3e0d3269.
The reason is bad_opcode_err is really for unknown err which ends up to default branch, so it would be better to maintain its message as it is(which good software should be). Including case for OP_RESERVED in message is rather confusing, and integrating bad_opcode_err with disabled_opcode_err is even worse as each of responsibility is literally different.
However I still think it's just wrong to throw bad_opcode_err for OP_VER and its variants, as those are not "unknown opcode", but really disabled ones.
Just want to make conclusion for this sooner or later. If stauts quo is good enough, u can close this PR.

**Comment by achow101** [comment]
ACK 2d6530b4c5300a29d87fb7042c6cec0b08cc9a0f

**Review by ajtowns — COMMENTED** [review]
I'm still ~0 on whether this actually improves anything, but I don't believe it causes any potential problems.

**Comment by fanquake** [comment]
I agree. At a minimum, given the PR description is near impossible to follow, it should be re-written. Especailly given the commit only contains `feat: OP_VERIF and OP_VERNOTIF throw disabled error`.

**Comment by ChrisCho-H** [comment]
> Throw `DISABLED_OPCODE` error for op_ver and its variants as it's known but just disabled opcodes. Currently, interpreter throws `BAD_OPCODE` which contains the message of `"Opcode missing or not understood"`, which isn't accurate and confusing.

Apologize for poor description, I just re-write to make it simple and clear.
I agree that this one is practically doesn't affect any, as only a few will push OP_VER opcodes on stack.
It's such a minor fix, and when I just found out that OP_VER throws BAD_OPCODE error, I thought it's just wrong to throw unknown opcode error just because it's not unknown opcode, considering only the limited context of interpreter code itself.

However, as @maflcko mentioned, changing consensus code to fix trivial thing(changing error case to throw) might be wrong(although I don't think this won't make any problem). It's ok to close this PR if that is better. As a noob I'm not 100% sure about which direction is better, and think experts like you will make much better decision.

**Pull request closed without merging**